internal/oapiffi/ffi2swagger.go

```diff
@@ -0,0 +1,139 @@
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/getkin/kin-openapi/openapi3"
+	"github.com/hyperledger/firefly/internal/oapispec"
+	"github.com/hyperledger/firefly/pkg/fftypes"
+)
+
+type FFISwaggerGen interface {
+	Generate(ctx context.Context, baseURL string, api *fftypes.ContractAPI, ffi *fftypes.FFI) (*openapi3.T, error)
+}
+
+// ffiSwaggerGen generates OpenAPI3 (Swagger) definitions for FFIs
+type ffiSwaggerGen struct {
+}
+
+func NewFFISwaggerGen() FFISwaggerGen {
+	return &ffiSwaggerGen{}
+}
+
+func (og *ffiSwaggerGen) Generate(ctx context.Context, baseURL string, api *fftypes.ContractAPI, ffi *fftypes.FFI) (swagger *openapi3.T, err error) {
+	hasLocation := !api.Location.IsNil()
+
+	routes := []*oapispec.Route{}
+	for _, method := range ffi.Methods {
+		routes = og.addMethod(routes, method, hasLocation)
+	}
+	for _, event := range ffi.Events {
+		routes = og.addEvent(routes, event, hasLocation)
+	}
+
+	return oapispec.SwaggerGen(ctx, routes, &oapispec.SwaggerGenConfig{
+		Title:       ffi.Name,
+		Version:     ffi.Version,
+		Description: ffi.Description,
+		BaseURL:     baseURL,
+	}), nil
+}
+
+func (og *ffiSwaggerGen) addMethod(routes []*oapispec.Route, method *fftypes.FFIMethod, hasLocation bool) []*oapispec.Route {
+	return append(routes, &oapispec.Route{
+		Name:             fmt.Sprintf("invoke_%s", method.Pathname),
+		Path:             fmt.Sprintf("invoke/%s", method.Pathname),
+		Method:           http.MethodPost,
+		JSONInputValue:   func() interface{} { return &fftypes.InvokeContractRequest{} },
+		JSONInputSchema:  func(ctx context.Context) string { return invokeRequestJSON(&method.Params, hasLocation).String() },
+		JSONOutputValue:  func() interface{} { return &fftypes.JSONObject{} },
+		JSONOutputSchema: func(ctx context.Context) string { return ffiParamsJSON(&method.Returns).String() },
+		JSONOutputCodes:  []int{http.StatusOK},
+	})
+}
+
+func (og *ffiSwaggerGen) addEvent(routes []*oapispec.Route, event *fftypes.FFIEvent, hasLocation bool) []*oapispec.Route {
+	// If the API has a location specified, there are no fields left to specify in the request body.
+	// Instead of masking them all (which causes Swagger UI some issues), explicitly set the schema to an empty object.
+	var schema func(ctx context.Context) string
+	if hasLocation {
+		schema = func(ctx context.Context) string { return `{"type": "object"}` }
+	}
+	return append(routes, &oapispec.Route{
+		Name:            fmt.Sprintf("subscribe_%s", event.Pathname),
+		Path:            fmt.Sprintf("subscribe/%s", event.Pathname),
+		Method:          http.MethodPost,
+		JSONInputValue:  func() interface{} { return &fftypes.ContractSubscribeRequest{} },
+		JSONInputMask:   []string{"Interface", "Event"},
+		JSONInputSchema: schema,
+		JSONOutputValue: func() interface{} { return &fftypes.ContractSubscription{} },
+		JSONOutputCodes: []int{http.StatusOK},
+	})
+}
+
+func invokeRequestJSON(params *fftypes.FFIParams, hasLocation bool) *fftypes.JSONObject {
```

Comment:
Can we change the name of this function to `invokeRequestJSONSchema`, and maybe add a comment that:

```go
// Returns an `fftypes.JSONObject` that contains a JSON Schema
````